[quant][graphmode][refactor] Remove observers in the end

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #31408 [quant][graphmode] `insert_quant_dequant` pass support shared class types
* **#31407 [quant][graphmode][refactor] Remove observers in the end**
* #31406 [quant][graphmode][refactor] Call `getQSchemeAndQParamMap` later in `quantizeTensors`

Summary:
Remove observers in the end instead of before quantize tensor
since we still need them to find the quantization paramters for each module instance

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19162367](https://our.internmc.facebook.com/intern/diff/D19162367)

jerryzh168 committed:
[quant][graphmode][refactor] Remove observers in the end

Summary:
Remove observers in the end instead of before quantize tensor
since we still need them to find the quantization paramters for each module instance

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Remove observers in the end"


Summary:
Remove observers in the end instead of before quantize tensor
since we still need them to find the quantization paramters for each module instance

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Remove observers in the end"


Summary:
Remove observers in the end instead of before quantize tensor
since we still need them to find the quantization paramters for each module instance

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 43d0c12:
* **2/2** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=43d0c1299d7dc960f8c66b7f3c9140d2c670fb40).

### :detective: 2 new failures recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3997399) pytorch_linux_xenial_cuda9_cudnn7_py3_build (1/2)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=76340652) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=76340652))
<details>
<summary>
<code>Dec 18 02:56:11 caused by: Connection refused (os error 111)</code>
</summary>

```
Dec 18 02:56:11 +++ eval 'extract_trap_cmd ' 
Dec 18 02:56:11 ++++ extract_trap_cmd 
Dec 18 02:56:11 ++++ printf '%s\n' '' 
Dec 18 02:56:11 +++ printf '%s\n' cleanup 
Dec 18 02:56:11 ++ trap -- ' 
Dec 18 02:56:11 cleanup' EXIT 
Dec 18 02:56:11 ++ which sccache 
Dec 18 02:56:11 ++ sccache --stop-server 
Dec 18 02:56:11 Stopping sccache server... 
Dec 18 02:56:11 error: couldn't connect to server 
Dec 18 02:56:11 caused by: Connection refused (os error 111) 
Dec 18 02:56:11 ++ true 
Dec 18 02:56:11 ++ rm /var/lib/jenkins/sccache_error.log 
Dec 18 02:56:11 rm: cannot remove '/var/lib/jenkins/sccache_error.log': No such file or directory 
Dec 18 02:56:11 ++ true 
Dec 18 02:56:11 ++ SCCACHE_ERROR_LOG=/var/lib/jenkins/sccache_error.log 
Dec 18 02:56:11 ++ SCCACHE_IDLE_TIMEOUT=1200 
Dec 18 02:56:11 ++ RUST_LOG=sccache::server=error 
Dec 18 02:56:11 ++ sccache --start-server 
Dec 18 02:56:11 Starting sccache server... 
Dec 18 02:56:11 ++ sccache --zero-stats 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3997302) pytorch_libtorch_linux_xenial_cuda9_cudnn7_py3_build (2/2)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=76341147) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=76341147))
<details>
<summary>
<code>Dec 18 02:57:12 caused by: Connection refused (os error 111)</code>
</summary>

```
Dec 18 02:57:12 +++ eval 'extract_trap_cmd ' 
Dec 18 02:57:12 ++++ extract_trap_cmd 
Dec 18 02:57:12 ++++ printf '%s\n' '' 
Dec 18 02:57:12 +++ printf '%s\n' cleanup 
Dec 18 02:57:12 ++ trap -- ' 
Dec 18 02:57:12 cleanup' EXIT 
Dec 18 02:57:12 ++ which sccache 
Dec 18 02:57:12 ++ sccache --stop-server 
Dec 18 02:57:12 Stopping sccache server... 
Dec 18 02:57:12 error: couldn't connect to server 
Dec 18 02:57:12 caused by: Connection refused (os error 111) 
Dec 18 02:57:12 ++ true 
Dec 18 02:57:12 ++ rm /var/lib/jenkins/sccache_error.log 
Dec 18 02:57:12 rm: cannot remove '/var/lib/jenkins/sccache_error.log': No such file or directory 
Dec 18 02:57:12 ++ true 
Dec 18 02:57:12 ++ SCCACHE_ERROR_LOG=/var/lib/jenkins/sccache_error.log 
Dec 18 02:57:12 ++ SCCACHE_IDLE_TIMEOUT=1200 
Dec 18 02:57:12 ++ RUST_LOG=sccache::server=error 
Dec 18 02:57:12 ++ sccache --start-server 
Dec 18 02:57:12 Starting sccache server... 
Dec 18 02:57:12 ++ sccache --zero-stats 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Remove observers in the end"


Summary:
Remove observers in the end instead of before quantize tensor
since we still need them to find the quantization paramters for each module instance

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM reviewed (commented):
I think this PR has something to do with my comment to the previous PR. Probably, this one becomes unnecessary too if you address my remarks to the previous one.

jerryzh168 commented:
> I think this PR has something to do with my comment to the previous PR. Probably, this one becomes unnecessary too if you address my remarks to the previous one.

we need to delay the removal of observer nodes because this information is still needed for other module instances to find the qparams.

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Remove observers in the end"


Summary:
Remove observers in the end instead of before quantize tensor
since we still need them to find the quantization paramters for each module instance

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19162367](https://our.internmc.facebook.com/intern/diff/D19162367)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Remove observers in the end"


Summary:
Remove observers in the end instead of before quantize tensor
since we still need them to find the quantization paramters for each module instance

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19162367](https://our.internmc.facebook.com/intern/diff/D19162367)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -878,9 +890,6 @@ void InsertQuantDeQuantHelper::run(
-  GRAPH_DUMP("Before Quantize Tensors:", graph);
   Value* self = graph->inputs()[0];
   quantizeTensors(module, graph.get(), self);
   GRAPH_DUMP("After Quantize Tensors:", graph);
```
Why can't we just move them after this line? That should do the trick and we won't need another (module-level) function for removing observers. It feels strange to keep observer nodes around in the graph where we already inserted all q-dq nodes.

jerryzh168 replied:
we have to keep the observer node around since we need them in order to get the quantization parameters, if they are removed we have to find another way to get these parameters and we'll create two path to get these parameters, that's probably not a great trade off.

ZolotukhinM reviewed (approved):
Overall looks good.

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Remove observers in the end"


Summary:
Remove observers in the end instead of before quantize tensor
since we still need them to find the quantization paramters for each module instance

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19162367](https://our.internmc.facebook.com/intern/diff/D19162367)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Remove observers in the end"


Summary:
Remove observers in the end instead of before quantize tensor
since we still need them to find the quantization paramters for each module instance

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19162367](https://our.internmc.facebook.com/intern/diff/D19162367)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Remove observers in the end"


Summary:
Remove observers in the end instead of before quantize tensor
since we still need them to find the quantization paramters for each module instance

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19162367](https://our.internmc.facebook.com/intern/diff/D19162367)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Remove observers in the end"


Summary:
Remove observers in the end instead of before quantize tensor
since we still need them to find the quantization paramters for each module instance

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19162367](https://our.internmc.facebook.com/intern/diff/D19162367)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Remove observers in the end"


Summary:
Remove observers in the end instead of before quantize tensor
since we still need them to find the quantization paramters for each module instance

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19162367](https://our.internmc.facebook.com/intern/diff/D19162367)

[ghstack-poisoned]
